k8s/eks/README.md

```diff
@@ -0,0 +1,284 @@
+sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
+kubectl version --client
+```
+
+Specific version (required when your cluster is not compatible with the latest version):
+
+```
+curl -LO https://dl.k8s.io/release/v1.22.0/bin/linux/amd64/kubectl
+sudo install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
+kubectl version --client
+```
+
+**4. AWS CLI (v2) with AWS credentials and ECR login**
+
+First install the aws cli v2 (note that v1 is already installed if using amazon linux, so you will first need to remove the v1 and install v2):
+
+https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html#cliv2-linux-install
+
+Then configure the credentials:
+
+https://docs.aws.amazon.com/cli/latest/userguide/getting-started-quickstart.html
+
+Verify you can log into ECR using the set credentials:
+
+```
+aws ecr get-login-password --region $AWS_REGION | docker login --username AWS --password-stdin $ACCOUNT_ID.dkr.ecr.$AWS_REGION.amazonaws.com
+
+# or in this case since the images are moved to a public repo:
+
+aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws/n6b0k8i7
+```
+
+**5. eksctl**
+
+Official installation guide:
+
+https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html
+
+```
+curl --silent --location "https://github.com/weaveworks/eksctl/releases/latest/download/eksctl_$(uname -s)_amd64.tar.gz" | tar xz -C /tmp
+sudo mv /tmp/eksctl /usr/local/bin
+eksctl version
+```
+
+
+## Step by step installation guide
+
+1. Pull the infra repo from here:
+
+git clone https://github.com/testground/infra.git
+
+and the testground repo from here:
+
+git clone https://github.com/testground/testground
+
+2. You need to populate the `.env` file with the needed parameters before creating the cluster. It is located in `infra/k8s/eks/`
+
+3. Run the installation script `./testground_install.sh`; note that it requires the `.env` file to be populated in order to run properly. In case of any missing variables, the setup will fail and you will get an error message
+
+The script explained in short:
+
+- Takes approx. 15-20 minutes to provision a cluster with 2 worker nodegroups, along with all other necessary resources and cluster workloads
+- All installation steps have been divided into functions, located in the `eks/bash/functions.sh` script
+- The script calls functions one by one
+- Output of every step is logged into a master log file that can be retrieved and reviewed, or even tailed in real time
+- It can also be used to uninstall a provisioned cluster; one of the first steps in the script is checking for existing resources
+- The `functions.sh` file relies on the content of `eks/yaml` - all cluster resources are located inside this folder
+- If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script
```

Comment:
By default, we are creating 2 worker nodes per NG. If you want to initially scale that, you need to edit the `functions.sh` on line 44 (the `make_cluster_config` function that creates the NG yaml file)
Also, scaling explanation has been added to the README since you left this comment.

Reply:
I can't see where we set 2 worker nodes per NG in the place you mentioned. When you mention a specific file/line, please add a permalink to that place.

Could we make that setting configurable through env?

Reply:
So the `.env` file already has this capability, and it is called `DESIRED_CAPACIY_INFRA=2` and `DESIRED_CAPACIY_PLAN=2`, which is then used in the `make_cluster_config` function (listed entire function below so you can find it easily:

```
make_cluster_config(){
    cat <<EOT >> ./$CLUSTER_NAME.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: $CLUSTER_NAME
  region: $REGION
managedNodeGroups:
  - name: ng-1-infra
    labels:
      "testground.node.role.infra": "true"
    instanceType: $INSTANCE_TYPE_INFRA
    ami: $AMI_INFRA
    desiredCapacity: $DESIRED_CAPACITY_INFRA
    volumeSize: $VOLUME_SIZE_INFRA
    privateNetworking: false
    availabilityZones: ["$AVAILABILITY_ZONE"]
    # ssh:
    #   allow: true
    #   # key needs to exist
    #   publicKeyPath: $SSH_PATH_INFRA
    iam:
      attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
        - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
   
    overrideBootstrapCommand: |
      #!/bin/bash
      /etc/eks/bootstrap.sh $CLUSTER_NAME \
        --kubelet-extra-args '--max-pods=58' \
        --use-max-pods false
  - name: ng-2-plan
    labels:
      "testground.node.role.plan": "true"
    instanceType: $INSTANCE_TYPE_PLAN 
    # Amazon EKS optimized Amazon Linux 2 v1.22 built on 08 Aug 2022
    ami: $AMI_PLAN
    desiredCapacity: $DESIRED_CAPACITY_PLAN
    volumeSize: $VOLUME_SIZE_PLAN
    privateNetworking: false
    availabilityZones: [$AVAILABILITY_ZONE]
    # ssh:
    #   allow: true
    #   publicKeyPath: $SSH_PATH_PLAN
    iam:
      attachPolicyARNs:
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryFullAccess
        - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
        - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
        - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
        - arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
    preBootstrapCommands:
      - "echo -e 'fs.file-max=3178504\nnet.core.somaxconn = 131072\nnet.netfilter.nf_conntrack_max = 1048576\nnet.core.netdev_max_backlog = 524288\nnet.core.rmem_max = 16777216\nnet.core.wmem_max = 16777216\nnet.ipv4.tcp_rmem = 16384 131072 16777216\nnet.ipv4.tcp_wmem = 16384 131072 16777216\nnet.ipv4.tcp_mem = 262144 524288 1572864\nnet.ipv4.tcp_max_syn_backlog = 131072\nnet.ipv4.ip_local_port_range = 10000 65535\nnet.ipv4.tcp_tw_reuse = 1\nnet.ipv4.ip_forward = 1\nnet.ipv4.conf.all.rp_filter = 0\nnet.ipv4.neigh.default.gc_thresh2 = 4096\nnet.ipv4.neigh.default.gc_thresh3 = 32768' >> /tmp/999-testground.conf; sudo cp /tmp/999-testground.conf /etc/sysctl.d/999-testground.conf; sudo sysctl -p /etc/sysctl.d/999-testground.conf; echo -e '*  soft  nproc  131072\n*  hard  nproc  262144\n*  soft  nofile 131072\n*  hard  nofile 262144' >> /tmp/999-limits.conf; sudo cp /tmp/999-limits.conf /etc/security/limits.d/999-limits.conf"
    # 234 is the max number of pods for c5.4xlarge
    overrideBootstrapCommand: |
      #!/bin/bash
      /etc/eks/bootstrap.sh $CLUSTER_NAME \
       --kubelet-extra-args '--max-pods=234' \
       --use-max-pods false
EOT
}
```

Reply:
So we only have to modify env and `functions.sh` stays in tact, right? If that's the case, let's update the note in the README then.

Reply:
Yes, exactly. It was intended to have only one config file that you need to edit, which is `.env`. Everything else runs with the parameters you specify in that file.
Makes things easier and simpler to have only one entrypoint to the script.

Reply:
But the README still says
```
- If you need to scale the worker nodegroups, or introduce any other changes, you are able to edit the `functions.sh` script
```

Reply:
Yeah, that is a mistake and will be corrected. It should say `.env` instead of `functions.sh`

Reply:
Why don't we do it straight away? Is there any reason why we're putting changes like that off?
```suggestion
- If you need to scale worker nodegroups, modify `DESIRED_CAPACIY_INFRA` or `DESIRED_CAPACIY_PLAN` in the environment file (`.env`)
```

Reply:
Because there are many more edits for the README file that I need to introduce, and I don't want to do it without checking some other things first.
Also, could you please edit your proposed change as there are 2 typos in it - should be `CAPACITY` instead of `CAPACIY`. 
Thanks.

Reply:
If you see something off there, just do a counter proposal. That's what this feature is for.

Reply:
Gathering all the changes is one huge blob is what got us here in the first place. Doing more smaller changes reduces review burden massively.